File: Day5/solution.ex
defmodule Solution do
  @stacks %{
    1 => 'PDQRVBHF',
    2 => 'VWQZDL',
    3 => 'CPRGQZLH',
    4 => 'BVJFHDR',
    5 => 'CLWZ',
    6 => 'MVGTNPRJ',
    7 => 'SBMVLRJ',
    8 => 'JPD',
    9 => 'VWNCD'
  }

  def parse_instruction(instruction) do
    Regex.scan(~r/move (.*) from (.*) to (.*)/, instruction, capture: :all_but_first)
    |> List.first()
    |> Enum.map(&String.to_integer/1)
  end

  def input(filename) do
    [_, instructions] = File.read!(filename) |> String.split("\n\n")

    instructions
    |> String.split("\n", trim: true)
    |> Enum.map(&parse_instruction/1)
  end

  def pop_count(stack, count), do: {Enum.take(stack, count), Enum.drop(stack, count)}

  def move_crates([], stack, _), do: Map.values(stack) |> Enum.map(&List.first/1) |> List.to_string()
  def move_crates([[count, from, to] | rest], stack, order) do
    {deleted, new} = pop_count(stack[from], count)

    new_stack = Map.replace!(stack, from, new) |> Map.replace(to, order.(deleted) ++ stack[to])
    move_crates(rest, new_stack, order)
  end

  def part_one(instructions), do: move_crates(instructions, @stacks, &Enum.reverse/1)
  def part_two(instructions), do: move_crates(instructions, @stacks, &Quark.id/1)

  def main do
    instructions = input("input.txt")
    IO.puts("Part one: #{part_one(instructions)}")
    IO.puts("Part two: #{part_two(instructions)}")
  end
end

File: Day5/solution.cpp
#include <array>
#include <filesystem>
#include <vector>
#include <fstream>
#include <fmt/format.h>
#include <ranges>
#include <string_view>
#include <algorithm>
#include <scn/scn.h>

namespace stdfs = std::filesystem;
namespace stdr = std::ranges;

using namespace std::string_view_literals;

std::string input(stdfs::path const& path) {
    std::ifstream reader(path);
    std::string text{std::istreambuf_iterator<char>(reader), std::istreambuf_iterator<char>()};

    return text.substr(text.find("\n\n") + 2);
}

std::string move_crates(std::array<std::string, 9> crates, std::string_view instructions, bool reverse = true) {
    for (auto line : instructions | std::views::split("\n"sv)) {
        int count, from, to;
        std::ignore = scn::scan(line, "move {} from {} to {}"sv, count, from, to);

        auto x = crates[from - 1].substr(0, count);
        if (reverse)
            stdr::reverse(x);

        crates[to - 1] = x + crates[to - 1];
        crates[from - 1].erase(0, count);
    }

    auto it = std::views::transform(crates, [](auto&& s) { return s[0]; });
    return std::string(it.begin(), it.end());
}

int main() {
    std::array<std::string, 9> crates {
        "PDQRVBHF",
        "VWQZDL",
        "CPRGQZLH",
        "BVJFHDR",
        "CLWZ",
        "MVGTNPRJ",
        "SBMVLRJ",
        "JPD",
        "VWNCD"
    };

    auto instr = input(stdfs::current_path().parent_path() / "input.txt");
    fmt::print("Part one: {}\n", move_crates(crates, instr));
    fmt::print("Part two: {}\n", move_crates(crates, instr, false));
}